### PR TITLE
fix inspect: return defargs[0] if obj.__dict__ is a property and raises exception

### PR DESCRIPTION
**Pull request by LeoHuckvale**
The fallback implemented in #2731 cannot return `obj.__dict__[name]` if the `__dict__` method has been redefined in such a way that it raises an exception when trying to access it.

This pull request adds a try-except block to work around this.

Any suggestions for a better implementation welcome!

### CODE REVIEW

**Comment by LeoHuckvale** [comment]
@seanfarley - I would appreciate your input here. Does this seem a reasonable extension to your original fix in #2731? I added a couple of tests which I think cover your original issue there, and the further issue I and @bobh66 encountered.

**Comment by seanfarley** [comment]
Let me check locally. Minor nit: Maybe it makes more sense to narrow the scope and return in the `try`,

``` python
try:
    return obj.__dict__[name]
except AttributeError:
    pass
```

Would `AttributeError` be enough?

**Comment by seanfarley** [comment]
It would seem:

``` python
try:
    return obj.__dict__[name]
except Exception:
    pass
```

would pass tests but I'm ok with the code as-is.

**Comment by seanfarley** [comment]
(in case it wasn't clear, this pull request works for my example)

**Comment by bobh66** [comment]
This appears to fix my scenario as well.

File: sphinx/util/inspect.py [hunk 1]
```diff
@@ -110,12 +110,23 @@ def safe_getattr(obj, name, *defargs):
...
         # NotImplementedError), so let's try to read the attribute directly
-        if name in obj.__dict__:
+
+        # We should also be aware that if `__dict__` has been overridden,
+        # this may also raise an exception.
+
+        try:
+            obj_dict = obj.__dict__
+        except Exception:
+            obj_dict = {}
```
Comment (tk0miya) [inline sphinx/util/inspect.py]: why is this line needed?
I think following code is much simpler:

```
try:
    if name in obj.__dict__:
        return obj.__dict__[name]
except Exception:
    # if __dict__ has been overridden...
    pass
```

**Comment by tk0miya** [comment]
Thanks, I've overlooked the case.
LGTM with nits.

File: sphinx/util/inspect.py (hunk 1 above)
Comment (LeoHuckvale) [inline sphinx/util/inspect.py]: I wanted to be very specific about where I was passing the exception, as it's such a broad `except` clause. Perhaps this is clearer:

``` python
try:
    # In case `__dict__` has been redefined as a `@property` such that
    # it may raise an error on access
    obj_dict = obj.__dict__
except Exception:
    pass
else:
    # We don't want to catch any exceptions here
    if name in obj_dict:
        return obj_dict[name]
```

**Comment by LeoHuckvale** [comment]
Thanks all for feedback - much appreciated.

@seanfarley - catching `AttributeError` isn't enough, apparently. The specific problem I and @bobh66 encountered was caused by [werkzeug.local.LocalProxy](https://github.com/pallets/werkzeug/blob/0.11.10/werkzeug/local.py), which does some weird stuff to attribute access and raises a `RuntimeError` when trying to get `__dict__` from the instance.

I'm currently looking into this a bit deeper to find if there's a more specific condition we can check - I'm not wholly comfortable with catching _any_ exception, but it might be the only way to do it in this instance.

**Comment by LeoHuckvale** [comment]
I've refactored to try to make the code clearer. I've also added comments to justify the broad exception.

**Comment by bobh66** [comment]
This version also works for my scenario.  Thanks.

**Comment by seanfarley** [comment]
Random question: why use `try-except-else` instead of just `try-except`? Do we really not want to catch an error in the `else` clause?

**Comment by tk0miya** [comment]
+1 for simple `try-except`.

**Comment by LeoHuckvale** [comment]
I've simplified it. My reasoning before was that, as we were catching _any_ exception, we should at least be as specific as possible about where we catch it. Broad `except` clauses are generally bad practice.

**Comment by tk0miya** [comment]
Thank you for updating. LGTM!
I think the comments above `except` clause describes enoughly why we use broad `except`.

**Comment by tk0miya** [comment]
Merged. Thank you for contribution!

**Comment by LeoHuckvale** [comment]
Thank you @tk0miya . Thanks all for your input. :)
